### PR TITLE
Shop billing data as resource

### PR DESCRIPTION
**Pull request by SirDomin**

### CODE REVIEW

File: src/Factory/ShopBillingDataFactory.php [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\RefundPlugin\Factory;
+
+use Sylius\Component\Resource\Factory\FactoryInterface;
+use Sylius\RefundPlugin\Entity\ShopBillingDataInterface;
+
+class ShopBillingDataFactory implements ShopBillingDataFactoryInterface
```
Comment (lchrusciel) [inline src/Factory/ShopBillingDataFactory.php:19]: could be finalized 
```suggestion
final class ShopBillingDataFactory implements ShopBillingDataFactoryInterface
```

**Review by lchrusciel — APPROVED** [review]

**Review by Tomanhez — APPROVED** [review]

File: src/Entity/ShopBillingData.php [hunk 2]
```diff
@@ -37,19 +37,73 @@ class ShopBillingData implements ShopBillingDataInterface
...
-        ?string $street,
-        ?string $city,
-        ?string $postcode
-    ) {
+    public function getId(): ?int
+    {
+        return $this->id;
+    }
+
+    public function setId(?int $id): void
```
Comment (GSadee) [inline src/Entity/ShopBillingData.php:45]: Similar to `CustomerBillingData`, remove the type of id not to force integers
```suggestion
    public function getId()
    {
        return $this->id;
    }

    public function setId($id): void
```

File: src/Entity/ShopBillingDataInterface.php [hunk 3]
```diff
@@ -13,19 +13,35 @@
 
 namespace Sylius\RefundPlugin\Entity;
 
-interface ShopBillingDataInterface
+use Sylius\Component\Resource\Model\ResourceInterface;
+
+interface ShopBillingDataInterface extends ResourceInterface
 {
-    public function id(): ?int;
```
Comment (GSadee) [inline src/Entity/ShopBillingDataInterface.php:18]: Deprecate this method instead of removing it now

File: src/Entity/ShopBillingDataInterface.php [hunk 4]
```diff
@@ -13,19 +13,35 @@
...
+    public function getCity(): ?string;
 
-    public function street(): ?string;
+    public function setCity(?string $city): void;
 
-    public function city(): ?string;
+    public function getPostcode(): ?string;
 
-    public function postcode(): ?string;
+    public function setPostcode(?string $postcode): void;
```
Comment (GSadee) [inline src/Entity/ShopBillingDataInterface.php:46]: As I see, you didn't remove these methods in `ShopBillingData` class, so why did you change them here? I would be for deprecating them instead of removing

File: src/Resources/config/services/generators.xml [hunk 5]
```diff
@@ -34,6 +35,16 @@
             <argument>@SyliusRefundPlugin/Download/creditMemo.html.twig</argument>
             <argument>%sylius.refund.template.logo_file%</argument>
         </service>
+
+        <service
+                id="Sylius\RefundPlugin\Factory\ShopBillingDataFactory"
+                decorates="sylius_refund.factory.shop_billing_data"
+                decoration-priority="256"
+                public="false"
```
Comment (GSadee) [inline src/Resources/config/services/generators.xml:43]: Wrong indentation

File: UPGRADE.md [hunk 6]
```diff
@@ -23,6 +23,23 @@
...
+
+1. The constructor of `Sylius\RefundPlugin\Generator\CreditMemoGenerator` has been changed:
+
+    ```diff
+        public function __construct(
+            LineItemsConverterInterface $lineItemsConverter,
+            LineItemsConverterInterface $shipmentLineItemsConverter,
+            TaxItemsGeneratorInterface $taxItemsGenerator,
+            CreditMemoFactoryInterface $creditMemoFactory,
+    +       ShopBillingDataFactoryInterface $shopBillingDataFactory
```
Comment (arti0090) [inline UPGRADE.md:37]: You need to change it after rebase 😉

File: src/Entity/ShopBillingDataInterface.php (hunk 3 above)
Comment (arti0090) [reply src/Entity/ShopBillingDataInterface.php:18]: This might come in handy 🙂  : https://github.com/Sylius/RefundPlugin/pull/313/files#diff-830a3324f5d10ec569eaf1c45a46c3859d022d2980c3ce2dd538ff4c5555bd14R20

Comment (arti0090) [reply src/Entity/ShopBillingDataInterface.php:18]: You can also check in PR I pasted how to suppress phpstan errors that will popup. See `phpstan.neon` file.

File: spec/Entity/ShopBillingDataSpec.php [hunk 7]
```diff
@@ -18,48 +18,49 @@
...
         $this->shouldImplement(ShopBillingDataInterface::class);
     }
 
     public function it_has_no_id_by_default(): void
     {
-        $this->id()->shouldReturn(null);
+        $this->getId()->shouldReturn(null);
     }
 
     public function it_has_company(): void
```
Comment (arti0090) [inline spec/Entity/ShopBillingDataSpec.php:31]: This is not a blocker for this PR but I would encourage you to add `a` or `an` articles before param.
F.e: `it_has_company` -> `it_has_a_company`

File: src/Generator/CreditMemoGenerator.php [hunk 8]
```diff
@@ -117,7 +123,7 @@ private function getToAddress(?ChannelShopBillingData $channelShopBillingData):
             return null;
         }
 
-        return new ShopBillingData(
+        return $this->shopBillingDataFactory->createWithData(
```
Comment (arti0090) [inline src/Generator/CreditMemoGenerator.php:126]: Maybe a method in factory called `createWithChannelShopBillingData` ? 😉 
See: https://github.com/Sylius/RefundPlugin/pull/313/files#diff-dc7a9ca1a71645aa3eb35cf2d342ab0543241c3b0173c6fe7d6ddf42a641be35R65

It won't block a merge imo - just a suggestion.

File: src/Entity/ShopBillingDataInterface.php (hunk 4 above)
Comment (SirDomin) [reply src/Entity/ShopBillingDataInterface.php]: i got them back along with deprecation info

File: spec/Entity/ShopBillingDataSpec.php (hunk 7 above)
Comment (SirDomin) [reply spec/Entity/ShopBillingDataSpec.php]: yeah, right

File: src/Generator/CreditMemoGenerator.php (hunk 8 above)
Comment (SirDomin) [reply src/Generator/CreditMemoGenerator.php:115]: i would leave it as it is now

File: src/Resources/config/services/generators.xml [hunk 9]
```diff
@@ -35,6 +37,16 @@
...
         </service>
+
+        <service
+            id="Sylius\RefundPlugin\Factory\ShopBillingDataFactory"
+            decorates="sylius_refund.factory.shop_billing_data"
+            decoration-priority="256"
+            public="false"
+        >
+            <argument type="service" id="Sylius\RefundPlugin\Factory\ShopBillingDataFactory.inner" />
+        </service>
```
Comment (Zales0123) [inline src/Resources/config/services/generators.xml:48]: Ok, that's why we do not specify a factory in `config.yml` 🤔 Maybe we should skip the decoration? We can do a factory + entry in resource configuration, the same way as it's done for `RefundPayment` (and it would keep an application consistency 💃 )

**Review by Zales0123 — COMMENTED** [review]
Removal of `tests/Application/var/.gitignore` is not relevant in the scope of this PR 🖖

File: src/Resources/config/app/config.yml [hunk 10]
```diff
@@ -22,6 +22,9 @@ sylius_resource:
         sylius_refund.customer_billing_data:
             classes:
                 model: Sylius\RefundPlugin\Entity\CustomerBillingData
+        sylius_refund.shop_billing_data:
+            classes:
+                model: Sylius\RefundPlugin\Entity\ShopBillingData
```
[the remark this answers is not in the document]
Comment (arti0090) [reply src/Resources/config/app/config.yml:27]: @Zales0123 it makes some automagic and will lead to errors. I had some troubles when I left `factory` here. Maybe @lchrusciel could say something more as he helped me finding issue with it

File: src/Resources/config/services/generators.xml (hunk 9 above)
Comment (arti0090) [reply src/Resources/config/services/generators.xml:48]: Well I leaved comment in discussion above, but thing about specify a factory was something with this.

File: src/Resources/config/app/config.yml (hunk 10 above)
[the remark this answers is not in the document]
Comment (Zales0123) [reply src/Resources/config/app/config.yml:27]: My problem is, we're actually creating a resource, so **conceptually** it's perfectly reasonable to implement a _default_ factory here, which then can be decorated in the final application 🚀 If it does not work, then wake up, Sylius Team - we have a bug to fix 🤖

File: src/Entity/ShopBillingDataInterface.php [hunk 11]
```diff
@@ -13,19 +13,56 @@
...
 
-interface ShopBillingDataInterface
+use Sylius\Component\Resource\Model\ResourceInterface;
+
+interface ShopBillingDataInterface extends ResourceInterface
 {
-    public function id(): ?int;
+    public function getId();
+
+    public function setId($id): void;
```
Comment (pamil) [inline src/Entity/ShopBillingDataInterface.php:22]: We won't need `setId` on the interface - see https://github.com/Sylius/RefundPlugin/pull/315/files#r653493427.

File: src/Entity/ShopBillingDataInterface.php [hunk 12]
```diff
@@ -13,19 +13,56 @@
 
 namespace Sylius\RefundPlugin\Entity;
 
-interface ShopBillingDataInterface
+use Sylius\Component\Resource\Model\ResourceInterface;
+
+interface ShopBillingDataInterface extends ResourceInterface
 {
-    public function id(): ?int;
+    public function getId();
```
Comment (pamil) [inline src/Entity/ShopBillingDataInterface.php:20]:
```suggestion
    /** @return mixed */
    public function getId();
```

File: phpstan.neon [hunk 13]
```diff
@@ -19,3 +19,9 @@ parameters:
         - '/Method Sylius\\RefundPlugin\\Entity\\CustomerBillingData::setId\(\) has parameter \$id with no typehint specified./'
         - '/Method Sylius\\RefundPlugin\\Entity\\CustomerBillingDataInterface::id\(\) has no return typehint specified./'
         - '/Method Sylius\\RefundPlugin\\Entity\\CustomerBillingDataInterface::setId\(\) has parameter \$id with no typehint specified./'
+        - '/Method Sylius\\RefundPlugin\\Entity\\ShopBillingData::id\(\) has no return typehint specified./'
```
Comment (pamil) [inline phpstan.neon:22]: We won't need those exclusions if we explicitly define return types as mixed.

File: src/Entity/ShopBillingDataInterface.php [hunk 14]
```diff
@@ -13,19 +13,56 @@
...
+
+    public function getCity(): ?string;
+
+    public function setCity(?string $city): void;
+
+    public function getPostcode(): ?string;
+
+    public function setPostcode(?string $postcode): void;
+
+    /** @deprecated this function is deprecated and will be removed in v1.0.0. Use CustomerBillingDataInterface::getId() instead */
```
Comment (pamil) [inline src/Entity/ShopBillingDataInterface.php:48]:
```suggestion
    /** 
     * @deprecated this function is deprecated and will be removed in v1.0.0. Use CustomerBillingDataInterface::getId() instead 
     * @see CustomerBillingDataInterface::getId
     * @return mixed
     */
```

File: src/Resources/config/services/generators.xml (hunk 9 above)
Comment (pamil) [reply src/Resources/config/services/generators.xml:48]: I'd try to use a factory here, it shouldn't cause any errors in this relatively easy use-case.

**Review by pamil — DISMISSED** [review]

File: src/Entity/ShopBillingData.php [hunk 15]
```diff
@@ -37,23 +37,77 @@ class ShopBillingData implements ShopBillingDataInterface
...
     protected $postcode;
 
-    public function __construct(
-        ?string $company,
-        ?string $taxId,
-        ?string $countryCode,
-        ?string $street,
-        ?string $city,
-        ?string $postcode
-    ) {
```
[the remark this answers is not in the document]
Comment (SirDomin) [reply src/Entity/ShopBillingData.php:47]: ShopBIllingDataFactory::create new is still supported, i dont know why should we throw unsupportedMethod?

File: UPGRADE.md [hunk 16]
```diff
@@ -26,6 +26,9 @@
 1. Constructor of `Sylius\RefundPlugin\Entity\CustomerBillingData` has been removed and now `CustomerBillingData` entity
    is created by `Sylius\RefundPlugin\Factory\CustomerBillingDataFactory`.
 
+1. Constructor of `Sylius\RefundPlugin\Entity\ShopBillingData` has been removed and now `ShopBillingData` entity
```
Comment (Zales0123) [inline UPGRADE.md:29]: `Constructor of `Sylius\RefundPlugin\Entity\ShopBillingData` has been removed` nope 💃

File: src/Factory/ShopBillingDataFactory.php [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+
+use Sylius\Component\Resource\Exception\UnsupportedMethodException;
+use Sylius\RefundPlugin\Entity\ShopBillingData;
+use Sylius\RefundPlugin\Entity\ShopBillingDataInterface;
+
+final class ShopBillingDataFactory implements ShopBillingDataFactoryInterface
+{
+    public function createNew(): ShopBillingDataInterface
+    {
+        throw new UnsupportedMethodException('createNew');
```
Comment (Zales0123) [inline src/Factory/ShopBillingDataFactory.php:24]: Exception message should be a little bit more explainable, something like: `To create ShopBillingData, you need to provide some arguments. Use ShopBillingDataFactoryInterface::createWithData instead`

File: src/Resources/config/app/config.yml (hunk 10 above)
[the remark this answers is not in the document]
Comment (Zales0123) [reply src/Resources/config/app/config.yml:27]: I suppose we're still missing this configuration?

**Review by Zales0123 — COMMENTED** [review]
`tests/Application/var/.gitignore` is still removed and it should not be 😄

File: src/Factory/ShopBillingDataFactory.php [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+
+    public function createWithData(
+        ?string $company,
+        ?string $taxId,
+        ?string $countryCode,
+        ?string $street,
+        ?string $city,
+        ?string $postcode
+    ): ShopBillingDataInterface {
+        $shopBillingData = new ShopBillingData();
```
Comment (GSadee) [inline src/Factory/ShopBillingDataFactory.php:35]: You should use here `className` similar to: https://github.com/Sylius/Sylius/blob/master/src/Sylius/Component/Order/Factory/OrderItemUnitFactory.php#L34 or go back to decorating the factory

File: src/Resources/config/app/config.yml (hunk 10 above)
[the remark this answers is not in the document]
Comment (GSadee) [reply src/Resources/config/app/config.yml:27]: I've searched the Sylius and there is used only in one place different factory that default one: https://github.com/Sylius/Sylius/blob/master/src/Sylius/Bundle/OrderBundle/DependencyInjection/Configuration.php#L110. All the others decorate the default one, so I'm not sure what is the recommended way to do. I would be probably for decorating, as it is more consistent and there is no problem with the type of created entity.

File: spec/Factory/ShopBillingDataFactorySpec.php [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+declare(strict_types=1);
+
+namespace spec\Sylius\RefundPlugin\Factory;
+
+use PhpSpec\ObjectBehavior;
+use Sylius\Component\Resource\Exception\UnsupportedMethodException;
+use Sylius\RefundPlugin\Entity\ShopBillingData;
+use Sylius\RefundPlugin\Factory\ShopBillingDataFactoryInterface;
+
+class ShopBillingDataFactorySpec extends ObjectBehavior
```
Comment (Zales0123) [inline spec/Factory/ShopBillingDataFactorySpec.php:21]: Missing final

File: src/Factory/ShopBillingDataFactory.php (hunk 18 above)
Comment (Zales0123) [reply src/Factory/ShopBillingDataFactory.php:35]: As `$className` is private property (😞) of a parent class, we should probably override a constructor and assign it to the property here as well. But yes, using this parameter is required 👌

File: src/Factory/ShopBillingDataFactoryInterface.php [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+declare(strict_types=1);
+
+namespace Sylius\RefundPlugin\Factory;
+
+use Sylius\Component\Resource\Factory\FactoryInterface;
+use Sylius\RefundPlugin\Entity\ShopBillingDataInterface;
+
+interface ShopBillingDataFactoryInterface extends FactoryInterface
+{
+    public function createNew(): ShopBillingDataInterface;
```
Comment (Zales0123) [inline src/Factory/ShopBillingDataFactoryInterface.php:21]: I believe we don't have to define this method, as we extend the `FactoryInterface`

File: src/Resources/config/app/config.yml (hunk 10 above)
[the remark this answers is not in the document]
Comment (Zales0123) [reply src/Resources/config/app/config.yml:27]: 👍 for consistency, even though I'm sad, as it would make more sense to use this configuration as a default approach 😄

[the remark this answers is not in the document]
Comment (SirDomin) [reply src/Resources/config/app/config.yml:27]: i got it back to solution with customer_billing_data

File: src/Factory/ShopBillingDataFactory.php (hunk 18 above)
Comment (SirDomin) [reply src/Factory/ShopBillingDataFactory.php]: its decorated now

File: src/Factory/ShopBillingDataFactoryInterface.php (hunk 20 above)
Comment (SirDomin) [reply src/Factory/ShopBillingDataFactoryInterface.php:21]: i reverted it to decorating

Comment (Zales0123) [reply src/Factory/ShopBillingDataFactoryInterface.php:21]: Still not needed, the method is in `FactoryInterface` nonetheless 😄

File: src/Resources/config/services/factories.xml [hunk 21]
```diff
@@ -31,6 +31,15 @@
             <argument type="service" id="Sylius\RefundPlugin\Provider\CurrentDateTimeImmutableProviderInterface" />
         </service>
 
+        <service
+                id="Sylius\RefundPlugin\Factory\ShopBillingDataFactory"
+                decorates="sylius_refund.factory.shop_billing_data"
+                decoration-priority="256"
+                public="false"
```
Comment (Zales0123) [inline src/Resources/config/services/factories.xml:38]: Too big indent 🖖

File: src/Factory/ShopBillingDataFactoryInterface.php (hunk 20 above)
Comment (SirDomin) [reply src/Factory/ShopBillingDataFactoryInterface.php]: yeah its not required to add it in this interface aswell

**Review by Zales0123 — APPROVED** [review]

**Comment by Zales0123** [comment]
Thank you, @SirDomin! :1st_place_medal:
